src/librustc_typeck/check/writeback.rs

```diff
@@ -374,8 +376,30 @@ impl<'cx, 'gcx, 'tcx> WritebackCx<'cx, 'gcx, 'tcx> {
             };
 
             self.tables
-                .user_provided_tys_mut()
+                .user_provided_types_mut()
                 .insert(hir_id, c_ty.clone());
+
+            if let ty::UserTypeAnnotation::TypeOf(_, user_substs) = c_ty.value {
+                if self.rustc_dump_user_substs {
+                    // This is a unit-testing mechanism.
+                    let node_id = self.tcx().hir().hir_to_node_id(hir_id);
+                    let span = self.tcx().hir().span(node_id);
+                    // We need to buffer the errors in order to guarantee a consistent
+                    // order when emitting them.
+                    let err = self.tcx().sess.struct_span_err(
+                        span,
+                        &format!("user substs: {:?}", user_substs)
+                    );
+                    err.buffer(&mut errors_buffer);
```

Comment:
@pnkfelix said (on a commit that since got changed in a rebase):
> Ugh this is going to be be yet another case of something where buffering upsets the use of tools like -Z treat-err-as-bug
> 
> Has we captured somewhere the pattern of buffering only if that and other related flags are off? Eg as a method on tcx? You don’t need to add such a method in this PR, but please double check if other spots that go buffering are using such a tcx method.

As far as I can gather, the only other place using diagnostic buffering like this is the borrow checker - where there are no such helpers. In this instance, buffering only takes place when the #[rustc_dump_user_substs] annotation is in use and that only happens in a handful of tests - it seems unlikely that it would interfere with usage of -Z treat-err-as-bug.

Reply:
Okay I must have misremembered. Chances are I tried adding such an abstraction and then later abandoned the attempt due to the code churn involved